crates/cxx-qt-lib/src/core/qurl.rs

```diff
@@ -63,21 +196,190 @@ pub struct QUrl {
-    //
-    // fragment: Option<String>,
-    // host: Option<String>,
-    // password: Option<String>,
-    // path: Option<String>,
-    // port: Option<u16>,
-    // query: Option<String>,
-    // scheme: Option<String>,
-    // userName: Option<String>,
+    /// Returns the authority of the URL if it is defined; otherwise an empty string is returned.
+    pub fn authority_or_default(&self) -> ffi::QString {
+        ffi::qurl_authority(self)
+    }
+
+    /// Returns the name of the file, excluding the directory path.
+    ///
+    /// Note that, if this QUrl object is given a path ending in a slash, the name of the file is considered empty.
+    ///
+    /// If the path doesn't contain any slash, it is fully returned as the fileName.
+    pub fn file_name(&self) -> ffi::QString {
+        ffi::qurl_file_name(self)
+    }
+
+    /// Returns the fragment of the URL.
+    pub fn fragment(&self) -> Option<ffi::QString> {
+        if self.has_fragment() {
+            Some(self.fragment_or_default())
+        } else {
+            None
+        }
+    }
+
+    /// Returns the fragment of the URL if it is defined; otherwise an empty string is returned.
+    pub fn fragment_or_default(&self) -> ffi::QString {
+        ffi::qurl_fragment(self)
+    }
+
+    /// Parses input and returns the corresponding QUrl. input is assumed to be in encoded form, containing only ASCII characters.
+    pub fn from_encoded(input: &ffi::QByteArray) -> Self {
+        ffi::qurl_from_encoded(input)
+    }
+
+    /// Returns a QUrl representation of localFile, interpreted as a local file.
+    /// This function accepts paths separated by slashes as well as the native separator for this platform.
+    pub fn from_local_file(local_file: &ffi::QString) -> Self {
+        ffi::qurl_from_local_file(local_file)
+    }
+
+    /// Returns a decoded copy of input. input is first decoded from percent encoding,
+    /// then converted from UTF-8 to unicode.
+    pub fn from_percent_encoding(input: &ffi::QByteArray) -> ffi::QString {
+        ffi::qurl_from_percent_encoding(input)
+    }
+
+    /// Returns a valid URL from a user supplied userInput string if one can be deduced.
+    /// In the case that is not possible, an invalid QUrl() is returned.
+    pub fn from_user_input(user_input: &ffi::QString, working_directory: &ffi::QString) -> Self {
+        ffi::qurl_from_user_input(user_input, working_directory)
+    }
+
+    /// Returns the host of the URL if it is defined; otherwise an empty string is returned.
+    pub fn host_or_default(&self) -> ffi::QString {
+        ffi::qurl_host(self)
+    }
+
+    /// Returns the current whitelist of top-level domains that are allowed to have non-ASCII characters in their compositions.
+    pub fn idn_whitelist() -> ffi::QStringList {
+        ffi::qurl_idn_whitelist()
+    }
+
+    /// Returns the password of the URL if it is defined; otherwise an empty string is returned.
+    pub fn password_or_default(&self) -> ffi::QString {
```

Comment:
Missing `fn password(&self) -> Option<ffi::QString>`

Reply:
So the problem is, `QUrl` doesn't have a `hasPassword()` or `isPassword()` method, the only way the docs say is an empty string. So we would have to do `password_or_default().is_empty() -> None`.

However my thought was, what if you have a password but the password is an empty string ? or the same with a username ? As null and empty are different but the `QUrl` API doesn't discriminate between these...

Again suggestions welcome.

Reply:
Alright, then stick with `password_or_default` for now.
Didn't check whether there was a `hasX` method available for these.

Reply:
tbh i think the `QUrl` API itself in Qt isn't great in this aspect as it doesn't consider null vs empty and is generally slightly inconsistent. I'll see which others i can fix though

Reply:
Hmm https://datatracker.ietf.org/doc/html/rfc3986#section-3.2.1 does mention about an empty password `unless the data after the colon is the empty string (indicating no password).`

Reply:
Then it seems there's no way to differentiate between an empty and an invalid password by Qts API, so let's just stick with `password_or_default`.